### DISCUSSION TITLE
how to get an image stream from the cameras?

### DISCUSSION THREAD

**Discussion by pattacini** [robotology/community#311] [Q&A] [answered]
@holygame commented on [Wed Jul 25 2018](https://github.com/robotology/icub-tutorials/issues/12)

I'm trying to get an image stream from the icub/icubSim cameras to use it with the ovrheadset driver .
If my understanding is correct , the ports where the images are streamed are  /icub/cam/left  and /icub/cam/right. 
i'm stuck on how to actually get the cameras to send images on those ports. For example , when controlling the head  we open the "remote_controlboard" device and connect it to the appropriate ports. What device and what interfaces do we need open to stream the output of the cameras ( for both icub and icubSim ).

Thank you for your help.

---

@pattacini commented on [Wed Jul 25 2018](https://github.com/robotology/icub-tutorials/issues/12#issuecomment-407764198)

Hi @holygame 
Welcome in robotology.

The proper place where discussing topics and asking for assistance is our Q&A repo. Therefore, I'll redirect there your request.

**Comment by barbalberto** [comment, marked as answer]
Hi @holygame,

the idea is the same, the device you have to use is the ` remote_grabber`. You can find the documentation here: http://www.yarp.it/classyarp_1_1dev_1_1RemoteFrameGrabber.html

The parameters required are:
"remote" "/imageSourcePort"
"local" "/myPort"

Also, you can find a list of available network wrapper with `yarpdev --list`

**Comment by holygame** [comment]
hi @barbalberto , thanks to you I was finally able to stream the images and see them through OpenCV!
Now I'm trying to stream them to the oculus headset.
the problem i'm facing is how to open the ovrheadset driver , there are 6 parameters that are required ;
- "tfLocal"
- "tfRemote"
- "tf_left_hand_frame"
- "tf_right_hand_frame"
- "tf_root_frame"
- "stick_as_axis"
- "gui_elements"

i couldn't find any documentation about those , could you please give me some hints ? 
thank you so much for the help!

**Comment by barbalberto** [comment]
I ping @aerydna who made that device to get more info about it (and didn't write the documentation!! :imp:  )

**Comment by aerydna** [comment]
documentation added via pr. 
https://github.com/robotology/yarp/pull/1821

**Comment by holygame** [comment]
I've tried different value for the parameters but I can't get the device to work.  I'm just interested in connecting the cameras  of the icub with the oculus  , what values of this parameters I should use ?

**Comment by barbalberto** [comment]
Please provide a description of the error you get when launching the device.

**Comment by holygame** [comment]
Could you please check if this default values make sense ? in the meantime i'll prepare the screenshot of the error 


 
```
* Parameters accepted in the config argument of the open method: 


 
* |   Parameter name      | Type   | Units | Default Value | Required  | Description                               | Notes | 


 
* |:---------------------:|:------:|:-----:|:-------------:|:---------:|:-----------------------------------------:|:-----:| 
* | tfLocal               | string |       |      /local/occulus         | yes       | local port name receiving and posting tf  |       | 
* | tfRemote              | string |       |     ???oculus???          | yes       | name of the transformServer port          |       | 
* | tf_left_hand_frame    | string |       |     right_hand          | Yes       | name of the left hand frame               |       | 
* | tf_right_hand_frame   | string |       |   left_hand            | yes       | name of the right hand frame              |       | 
* | tf_root_frame         | string |       |    torso           | yes       | name of the root frame                    |       | 
* | stick_as_axis         | bool   |       |       false        | yes       | if axes shoud be published as sticks      |       | 
* | gui_elements          | int    |       |       3        | yes       | number of the gui element to visualize    |     ```

**Comment by holygame** [comment]
```
||| clearing context
||| adding context [myModule/conf]
||| configuring
||| default config file specified as myModule.ini
||| checking [C:\Users\AnDy\Yarp_Projects\ICubCamToOculus\build\Release\myModule.ini] (pwd)
||| found C:\Users\AnDy\Yarp_Projects\ICubCamToOculus\build\Release\myModule.ini
yarp: Port / active at tcp://10.0.0.68:10055/
||| finding file [icubEyes[.ini]
IN|||F checking [C:\Users\AnDy\Yarp_Projects\ICubCamToOculus\build\Release\icubEyes.ini] (pwd)
O||| found C:]\Users\AnDy\Yarp_Projects\ICubCamToOculus\build\Release\icubEyes.ini
Listening to terminal (type "quit" to stop module).
[DEBUG]config is  (device remote_grabber) (local "/oculus/display/right:i") (remote "/icub_sim/Cam/right")
yarp: Port /oculus/display/right:i active at tcp://10.0.0.68:10057/
[INFO]connecting  /oculus/display/right:i  to  /icub_sim/Cam/right
[ERROR]cannot connect  /oculus/display/right:i  to  /icub_sim/Cam/right
[INFO]created device <remote_grabber>. See C++ class yarp::dev::RemoteFrameGrabber for documentation.
[ERROR]unable to dynamic cast device to IFrameTransform interface
yarpdev: ***ERROR*** driver <ovrheadset> was found but could not open
ovrheadset not ready
[INFO]RFModule failed to open.
```

here is the configuration file used 

```
device ovrheadset
tfDevice remote_grabber
tfLocal /oculus/display/right:i
tfRemote /icub_sim/Cam/right
tf_left_hand_frame  /icub_sim/arm/left
tf_right_hand_frame /icub_sim/arm/right
tf_root_frame str
stick_as_axis false
gui_elements 0
```

**Comment by barbalberto** [comment]
The errror says
`[ERROR]cannot connect  /oculus/display/right:i  to  /icub_sim/Cam/right`
so either the port producing the image is not open yet or its name is different.

Check the port name with `yarp name list`

**Comment by holygame** [comment]
I've tried with the correct port name but still have problems , it's also weird that i have to use one port for the images but i have 2 port streaming the images ( left and right eyes )

```
||| clearing context
||| adding context [myModule/conf]
||| configuring
||| default config file specified as myModule.ini
||| checking [C:\Users\AnDy\Yarp_Projects\ICubCamToOculus\build\Release\myModule.ini] (pwd)
||| found C:\Users\AnDy\Yarp_Projects\ICubCamToOculus\build\Release\myModule.ini
yarp: Port / active at tcp://10.0.0.68:10055/
||| finding file [icubEyes.ini]
[||| checking [CINFO:\Users\AnDy\Yarp_Projects\I]CubCaLimToOculus\build\Reslease\icubEyes.init] (pwd)
en||| found C:\Users\AnDy\Yiarp_Projects\ICubCamToOculus\build\Release\icubEyes.inin
g to terminal (type "quit" to stop module).
[DEBUG]config is  (device remote_grabber) (local "/oculus/display/right:i") (remote "/icubSim/cam/right")
yarp: Port /oculus/display/right:i active at tcp://10.0.0.68:10057/
[INFO]connecting  /oculus/display/right:i  to  /icubSim/cam/right
yarp: Receiving input from /icubSim/cam/right to /oculus/display/right:i using tcp
yarp: Sending output from /oculus/display/right:i to /icubSim/cam/right using tcp
[INFO]created device <remote_grabber>. See C++ class yarp::dev::RemoteFrameGrabber for documentation.
[ERROR]unable to dynamic cast device to IFrameTransform interface
yarpdev: ***ERROR*** driver <ovrheadset> was found but could not open
yarp: Removing input from /icubSim/cam/right to /oculus/display/right:i
yarp: Removing output from /oculus/display/right:i to /icubSim/cam/right
ovrheadset not ready
[INFO]RFModule failed to open.`

```
config

```
device ovrheadset
tfDevice remote_grabber
tfLocal /oculus/display/right:i
tfRemote /icubSim/cam/right
tf_left_hand_frame  /icubSim/arm/left
tf_right_hand_frame /icubSim/arm/right
tf_root_frame str
stick_as_axis false
gui_elements 0
```

**Comment by barbalberto** [comment]
The config file is not correct
`tfDevice remote_grabber`     tfDevice has to be the device handling the Transform Frames, while the `remote_grabber` is for the images.
@aerydna can you assist @holygame in creating a correct config file?

**Comment by aerydna** [comment]
hi,

device parameter is correct

`tfDevice` paramenter should specify the only client tf device currently avaiable in yarp: `transformClient`

tfLocal parameter will be the arbitrary name of the local tf port opened by the transform client. you can put any name as long for it to respect the yarp naming rules for ports

`tfRemote` should be equal to the name the name of the port opened by the `transformServer`. the `ovrdevice`, need the TransformServer to be up and running. you can open it with `yarpdev --device transformServer` and relative parameters.

tf_left/right_hand_frame and root_frame are the arbitrary names of the frames published by the ovr device. you can choose any name as long as them to be consistent with the frames searched by the client application.

the other parameters are fine

**Comment by holygame** [comment]
thank you @aerydna and @barbalberto it's working perfectly !
